File: lib/ai/mock.ts
import type {
  ClassificationResult,
  ExtractionResult,
  LanguageResult,
  RecommendationResult,
  RiskResult,
  ValidationResultAI,
} from "./schemas";
import type { DocumentType } from "../types";

// Deterministic, heuristic stand-ins for the Gemini agents. Used when no API
// key is configured (or FORCE_MOCK=1). They run real regex/keyword analysis on
// the OCR text, so the offline path produces meaningful — not random — output.

const KNOWN_AUTHORITIES = [
  "bundesagentur für arbeit",
  "ausländerbehörde",
  "auslanderbehorde",
  "bundesamt für migration",
  "home office",
  "immigration",
  "migration",
  "ministero dell'interno",
  "préfecture",
  "prefecture",
];

const TYPE_KEYWORDS: { type: DocumentType; words: string[] }[] = [
  { type: "Residence Permit", words: ["residence permit", "aufenthaltstitel", "aufenthaltserlaubnis", "residence card", "aufenthaltskarte"] },
  { type: "Work Permit", words: ["work permit", "arbeitserlaubnis", "beschäftigungserlaubnis", "work authorization", "permission to work"] },
  { type: "Employment Authorization", words: ["employment authorization", "ead", "arbeitsgenehmigung"] },
  { type: "Visa", words: ["visa", "visum", "schengen"] },
  { type: "Passport", words: ["passport", "reisepass", "passeport"] },
  { type: "National ID", words: ["national id", "identity card", "personalausweis", "carte d'identité"] },
  { type: "Driving License", words: ["driving licence", "driving license", "driver license", "führerschein", "permis de conduire"] },
];

function findDates(text: string): string[] {
  const iso = [...text.matchAll(/\b(\d{4})-(\d{2})-(\d{2})\b/g)].map((m) => m[0]);
  const dmy = [...text.matchAll(/\b(\d{1,2})[.\/](\d{1,2})[.\/](\d{2,4})\b/g)].map((m) => {
    const [, d, mo, y] = m;
    const yy = y.length === 2 ? `20${y}` : y;
    return `${yy}-${mo.padStart(2, "0")}-${d.padStart(2, "0")}`;
  });
  return [...iso, ...dmy];
}

export function mockClassify(text: string): ClassificationResult {
  const lower = text.toLowerCase();
  for (const { type, words } of TYPE_KEYWORDS) {
    if (words.some((w) => lower.includes(w))) {
      return { document_type: type, classification_confidence: 0.82, reasoning: "Keyword match (heuristic)" };
    }
  }
  if (lower.length < 30) return { document_type: "Unknown", classification_confidence: 0.3 };
  return { document_type: "Other", classification_confidence: 0.45 };
}

export function mockLanguage(text: string): LanguageResult {
  const lower = text.toLowerCase();
  const germanHints = ["und", "der", "die", "arbeit", "aufenthalt", "gültig", "ausländer", "behörde"];
  const isGerman = germanHints.filter((w) => lower.includes(w)).length >= 2;
  return { language: isGerman ? "German" : "English", is_english: !isGerman, english_translation: null };
}

export function mockExtract(text: string): ExtractionResult {
  const lower = text.toLowerCase();
  const dates = findDates(text).sort();
  const permitNo = text.match(/\b([A-Z]{2,3}(?:-[A-Z]{2,4})?-?\d{3,}[-\s]?\d*[A-Z0-9-]*)\b/)?.[1] ?? null;
  // Capture the value after "Name:" up to the next known field label (works on
  // both line-broken and merged single-line OCR/PDF text).
  const FIELD_BOUNDARY =
    "nationality|permit|country|issuing|issue|expiry|document|residence|employer|occupation|visa|date|status";
  const nameMatch =
    text.match(new RegExp(`(?:holder name|inhaber|\\bname|\\bnom)\\s*[:\\-]\\s*(.*?)\\s*(?:\\b(?:${FIELD_BOUNDARY})\\b|$)`, "i"))?.[1]?.trim() ||
    null;
  const authority = KNOWN_AUTHORITIES.find((a) => lower.includes(a)) ?? null;

  return {
    holder_name: nameMatch,
    nationality: null,
    permit_number: permitNo,
    permit_type: null,
    country: lower.includes("germany") || lower.includes("deutschland") ? "Germany" : null,
    issuing_authority: authority ? authority.replace(/\b\w/g, (c) => c.toUpperCase()) : null,
    issue_date: dates[0] ?? null,
    expiry_date: dates.length > 1 ? dates[dates.length - 1] : null,
    document_number: null,
    employer_restrictions: null,
    occupation_restrictions: null,
    visa_category: null,
    residence_status: lower.includes("authorized to work") || lower.includes("berechtigt") ? "Authorized to work" : null,
    notes: null,
  };
}

export function mockValidate(fields: ExtractionResult, ocrConfidence: number): ValidationResultAI {
  const present = (v: unknown) => v != null && String(v).trim().length > 0;
  const required = [fields.holder_name, fields.permit_number, fields.issuing_authority, fields.expiry_date];
  const fields_complete = required.every(present);
  const expiry_present = present(fields.expiry_date);
  const authority_recognized = present(fields.issuing_authority);
  const tampering: string[] = [];

  if (present(fields.issue_date) && present(fields.expiry_date)) {
    if (new Date(fields.expiry_date as string) < new Date(fields.issue_date as string)) {
      tampering.push("Expiry date precedes issue date");
    }
  }
  if (ocrConfidence < 0.6) tampering.push("Low OCR confidence");

  let status: ValidationResultAI["validation_status"] = "Valid";
  if (tampering.length) status = "Possibly Invalid";
  else if (!fields_complete || !authority_recognized) status = "Needs Review";

  const reasons: string[] = [];
  if (authority_recognized) reasons.push("Issuing authority present");
  else reasons.push("Issuing authority not found");
  if (expiry_present) reasons.push("Expiry date present");
  else reasons.push("Expiry date missing");
  if (fields_complete) reasons.push("Required fields present");
  else reasons.push("Some required fields missing");
  reasons.push(...tampering);

  return {
    validation_status: status,
    fields_complete,
    expiry_present,
    authority_recognized,
    format_matches: ocrConfidence >= 0.6 && fields_complete,
    tampering_indicators: tampering,
    reasons,
  };
}

export function mockRisk(validation: ValidationResultAI, ocrConfidence: number): RiskResult {
  let score = 10;
  const anomalies: string[] = [];
  if (validation.tampering_indicators.length) {
    score += 40 * validation.tampering_indicators.length;
    anomalies.push(...validation.tampering_indicators);
  }
  if (!validation.authority_recognized) {
    score += 20;
    anomalies.push("Unrecognized or missing issuing authority");
  }
  if (!validation.fields_complete) {
    score += 15;
    anomalies.push("Incomplete fields");
  }
  if (ocrConfidence < 0.6) {
    score += 15;
    anomalies.push("Low-quality scan");
  }
  return { risk_score: Math.max(0, Math.min(100, score)), anomalies, confidence: ocrConfidence };
}

export function mockRecommend(args: {
  validationStatus: string;
  riskScore: number;
  expired: boolean;
  fieldsComplete: boolean;
}): RecommendationResult {
  const reasons: string[] = [];
  let recommendation: RecommendationResult["recommendation"] = "Likely Valid";
  let requires = false;

  if (args.validationStatus === "Unknown") {
    recommendation = "Unable to Verify";
    requires = true;
    reasons.push("Document type or content could not be confirmed");
  } else if (args.riskScore >= 67 || args.validationStatus === "Possibly Invalid") {
    recommendation = "Needs Human Review";
    requires = true;
    reasons.push("Elevated authenticity risk or validation concerns");
  } else if (!args.fieldsComplete) {
    recommendation = "Document Incomplete";
    requires = true;
    reasons.push("Required fields are missing");
  } else if (args.expired) {
    recommendation = "Needs Human Review";
    requires = true;
    reasons.push("Document is expired");
  } else {
    recommendation = "Likely Valid";
    reasons.push("Recognized authority, complete fields, low risk");
  }

  const confidence = Math.max(0.3, Math.min(0.97, 1 - args.riskScore / 130));
  return { recommendation, reasons, requires_human_review: requires, confidence };
}

File: backend/app/ai/mock.py
"""Deterministic, heuristic stand-ins for the Gemini agents.

Used when no API key is configured (or FORCE_MOCK=1). They run real
regex/keyword analysis on the OCR text, so the offline path produces
meaningful - not random - output.
"""
from __future__ import annotations

import re
from typing import List

from .schemas import Classification, Extraction, Language, RecommendationResult, Risk, Validation

KNOWN_AUTHORITIES = [
    "bundesagentur für arbeit",
    "bundesagentur fur arbeit",
    "ausländerbehörde",
    "auslanderbehorde",
    "bundesamt für migration",
    "bundesamt fur migration",
    "home office",
    "immigration",
    "migration",
    "ministero dell'interno",
    "préfecture",
    "prefecture",
]

TYPE_KEYWORDS = [
    ("Residence Permit", ["residence permit", "aufenthaltstitel", "aufenthaltserlaubnis", "residence card", "aufenthaltskarte"]),
    ("Work Permit", ["work permit", "arbeitserlaubnis", "beschäftigungserlaubnis", "work authorization", "permission to work"]),
    ("Employment Authorization", ["employment authorization", "ead", "arbeitsgenehmigung"]),
    ("Visa", ["visa", "visum", "schengen"]),
    ("Passport", ["passport", "reisepass", "passeport"]),
    ("National ID", ["national id", "identity card", "personalausweis", "carte d'identité"]),
    ("Driving License", ["driving licence", "driving license", "driver license", "führerschein", "permis de conduire"]),
]


def _find_dates(text: str) -> List[str]:
    iso = re.findall(r"\b(\d{4})-(\d{2})-(\d{2})\b", text)
    out = [f"{y}-{m}-{d}" for (y, m, d) in iso]
    for d, mo, y in re.findall(r"\b(\d{1,2})[./](\d{1,2})[./](\d{2,4})\b", text):
        yy = f"20{y}" if len(y) == 2 else y
        out.append(f"{yy}-{mo.zfill(2)}-{d.zfill(2)}")
    return out


def mock_classify(text: str) -> Classification:
    low = text.lower()
    for doc_type, words in TYPE_KEYWORDS:
        if any(w in low for w in words):
            return Classification(document_type=doc_type, classification_confidence=0.82, reasoning="Keyword match (heuristic)")
    if len(low.strip()) < 30:
        return Classification(document_type="Unknown", classification_confidence=0.3)
    return Classification(document_type="Other", classification_confidence=0.45)


def mock_language(text: str) -> Language:
    low = text.lower()
    german_hints = ["und", "der", "die", "arbeit", "aufenthalt", "gültig", "ausländer", "behörde"]
    is_german = sum(1 for w in german_hints if w in low) >= 2
    return Language(language="German" if is_german else "English", is_english=not is_german, english_translation=None)


def mock_extract(text: str) -> Extraction:
    low = text.lower()
    dates = sorted(_find_dates(text))
    permit_match = re.search(r"\b([A-Z]{2,3}(?:-[A-Z]{2,4})?-?\d{3,}[-\s]?\d*[A-Z0-9-]*)\b", text)
    permit_no = permit_match.group(1) if permit_match else None

    boundary = "nationality|permit|country|issuing|issue|expiry|document|residence|employer|occupation|visa|date|status"
    name_match = re.search(
        rf"(?:holder name|inhaber|\bname|\bnom)\s*[:\-]\s*(.*?)\s*(?:\b(?:{boundary})\b|$)",
        text,
        re.IGNORECASE,
    )
    name = name_match.group(1).strip() if name_match and name_match.group(1).strip() else None

    authority = next((a for a in KNOWN_AUTHORITIES if a in low), None)

    return Extraction(
        holder_name=name,
        nationality=None,
        permit_number=permit_no,
        permit_type=None,
        country="Germany" if ("germany" in low or "deutschland" in low) else None,
        issuing_authority=authority.title() if authority else None,
        issue_date=dates[0] if dates else None,
        expiry_date=dates[-1] if len(dates) > 1 else None,
        document_number=None,
        employer_restrictions=None,
        occupation_restrictions=None,
        visa_category=None,
        residence_status=("Authorized to work" if ("authorized to work" in low or "berechtigt" in low) else None),
        notes=None,
    )


def mock_validate(fields: Extraction, ocr_confidence: float) -> Validation:
    def present(v) -> bool:
        return v is not None and len(str(v).strip()) > 0

    required = [fields.holder_name, fields.permit_number, fields.issuing_authority, fields.expiry_date]
    fields_complete = all(present(v) for v in required)
    expiry_present = present(fields.expiry_date)
    authority_recognized = present(fields.issuing_authority)
    tampering: List[str] = []

    if present(fields.issue_date) and present(fields.expiry_date):
        try:
            if fields.expiry_date < fields.issue_date:  # ISO strings compare lexicographically
                tampering.append("Expiry date precedes issue date")
        except TypeError:
            pass
    if ocr_confidence < 0.6:
        tampering.append("Low OCR confidence")

    if tampering:
        status = "Possibly Invalid"
    elif not fields_complete or not authority_recognized:
        status = "Needs Review"
    else:
        status = "Valid"

    reasons: List[str] = []
    reasons.append("Issuing authority present" if authority_recognized else "Issuing authority not found")
    reasons.append("Expiry date present" if expiry_present else "Expiry date missing")
    reasons.append("Required fields present" if fields_complete else "Some required fields missing")
    reasons.extend(tampering)

    return Validation(
        validation_status=status,
        fields_complete=fields_complete,
        expiry_present=expiry_present,
        authority_recognized=authority_recognized,
        format_matches=ocr_confidence >= 0.6 and fields_complete,
        tampering_indicators=tampering,
        reasons=reasons,
    )


def mock_risk(validation: Validation, ocr_confidence: float) -> Risk:
    score = 10.0
    anomalies: List[str] = []
    if validation.tampering_indicators:
        score += 40 * len(validation.tampering_indicators)
        anomalies.extend(validation.tampering_indicators)
    if not validation.authority_recognized:
        score += 20
        anomalies.append("Unrecognized or missing issuing authority")
    if not validation.fields_complete:
        score += 15
        anomalies.append("Incomplete fields")
    if ocr_confidence < 0.6:
        score += 15
        anomalies.append("Low-quality scan")
    return Risk(risk_score=max(0.0, min(100.0, score)), anomalies=anomalies, confidence=ocr_confidence)


def mock_recommend(validation_status: str, risk_score: float, expired: bool, fields_complete: bool) -> RecommendationResult:
    reasons: List[str] = []
    requires = False

    if validation_status == "Unknown":
        recommendation = "Unable to Verify"
        requires = True
        reasons.append("Document type or content could not be confirmed")
    elif risk_score >= 67 or validation_status == "Possibly Invalid":
        recommendation = "Needs Human Review"
        requires = True
        reasons.append("Elevated authenticity risk or validation concerns")
    elif not fields_complete:
        recommendation = "Document Incomplete"
        requires = True
        reasons.append("Required fields are missing")
    elif expired:
        recommendation = "Needs Human Review"
        requires = True
        reasons.append("Document is expired")
    else:
        recommendation = "Likely Valid"
        reasons.append("Recognized authority, complete fields, low risk")

    confidence = max(0.3, min(0.97, 1 - risk_score / 130))
    return RecommendationResult(
        recommendation=recommendation, reasons=reasons, requires_human_review=requires, confidence=confidence
    )
